src/open-response/components/runtime.tsx

```diff
@@ -29,17 +29,35 @@ export const Runtime: React.FC<IProps> = ({ authoredState, interactiveState, set
   const answerText = !interactiveState?.answerText ? authoredState.defaultAnswer : interactiveState.answerText;
   const attachedAudioFile = interactiveState?.audioFile ? interactiveState.audioFile : undefined;
   let recordedBlobs: Blob[] = [];
-  const [audioUrl, setAudioUrl] = React.useState<string | undefined>(undefined);
-  const [audioSupported, setAudioSupported] = React.useState(browserSupportsAudio);
-  const [recordingStarted, setRecordingStarted] = React.useState(false);
-  const [recordingDisabled, setRecordingDisabled] = React.useState(false);
-  const [recordingFailed, setRecordingFailed] = React.useState(false);
-  const [playDisabled, setPlayDisabled] = React.useState(false);
-  const [stopDisabled, setStopDisabled] = React.useState(true);
-  const mediaRecorderRef = React.useRef<MediaRecorder>();
-  const audioPlayerRef = React.useRef<HTMLAudioElement>(null);
-
-  React.useEffect(() => {
+  const [audioUrl, setAudioUrl] = useState<string | undefined>(undefined);
+  const [audioSupported, setAudioSupported] = useState(browserSupportsAudio);
+  const [recordingStarted, setRecordingStarted] = useState(false);
+  const [recordingDisabled, setRecordingDisabled] = useState(false);
+  const [recordingFailed, setRecordingFailed] = useState(false);
+  const [playDisabled, setPlayDisabled] = useState(false);
+  const [stopDisabled, setStopDisabled] = useState(true);
+  const mediaRecorderRef = useRef<MediaRecorder>();
+  const audioPlayerRef = useRef<HTMLAudioElement>(null);
+
+  useEffect(() => {
+    setOnUnload((options: IGetInteractiveState) => {
+      if (options.unloading) {
+        return new Promise<Record<string, unknown>>((resolve) => {
+          if (mediaRecorderRef.current) {
+            mediaRecorderRef.current.stop();
+            if (audioUrl) {
+              resolve;
+            }
+          } else {
+            resolve({});
+          }
```

Comment:
@dougmartin Thanks. This probably could be clearer, but line 47 -- `mediaRecorderRef.current.stop()` -- triggers a call to `handleAudioSave` which uses `setInteractiveState` to, as I understand it, take care of setting the final state. (See line 146.) After that, the `audioUrl` state variable is then set which allows the promise to resolve, or so I thought. It seems to work correctly in the browser anyway.

I had been following the example of `setOnUnload` in https://github.com/concord-consortium/question-interactives/blob/76a1e6a88f9a2877dab52a8e684aadf180d3b0aa/src/shared/components/iframe-runtime.tsx#L62 before you posted the other example on Slack. Does that seem potentially wrong, too? Or maybe it's doing something different that I missed? I see it's using that `resolveOnUnload` ref and is posting to iframe phone, but I don't think those add anything that we need here.

Also, if I change the promise type in line 45 to anything other than `any` or `Record<string, unknown>` and/or change `resolve` on line 49 to `resolve()` I get linting errors. If I remove the type assertion altogether and change `resolve` to `resolve({})` the linter is happy. Does that make sense? I'm afraid I'm missing something with how `resolve` might update the interactive state. I don't want it to since that's already taken care of by `handleAudioSave` being called.

Reply:
So this is working because the `stop()` method updating the interactive state resolves the onUnload handler in the host.  Can you try this which I think would be clearer?

```
  useEffect(() => {
    setOnUnload((options: IGetInteractiveState) => {
      if (options.unloading) {
        return new Promise(() => {
          // do not resolve to update the final interactive state, 
          // the stop() method will do that which will complete the onUnload promise in the host
          mediaRecorderRef.current?.stop();
        });
      }
      return Promise.resolve({});
    });
  }, [audioUrl]);
```

Reply:
@dougmartin That looks good but turns out to not resolve in cases where `mediaRecorderRef.current` isn't defined or when it is defined but the recording and saving process was allowed to finish.

My original code does handle those cases but is convoluted and doing other things wrong.

Would it make sense and help clarify things to add an `isDirty` state variable which is false by default, set to true when recording starts, and then set to false when the recording process completes? And then the `useEffect` would look something like the below?

`useEffect(() => {
    setOnUnload((options: IGetInteractiveState) => {
      const savedState = interactiveState ? interactiveState : {};
      if (options.unloading) {
        return new Promise(resolve => {
          if (isDirty) {
            // do not resolve to update the final interactive state, 
            // the stop() method will do that which will complete the onUnload promise in the host
            mediaRecorderRef.current?.stop();
          } else {
            resolve(savedState);
          }
        });
      }
      return Promise.resolve(savedState);
    });
 });`

Reply:
Or maybe `recordingInProgress` would be slightly better than `isDirty`?

Reply:
Good catch, I think you don't need the `isDirty` extra flag if you did this:

```
useEffect(() => {
    setOnUnload((options: IGetInteractiveState) => {
      if (options.unloading) {
        return new Promise(resolve => {
          if (mediaRecorderRef.current) {
            // do not resolve to update the final interactive state, 
            // the stop() method will do that which will complete the onUnload promise in the host
            mediaRecorderRef.current.stop();
          } else {
            resolve(interactiveState || {});
          }
        });
      }
      return Promise.resolve(interactiveState || {});
    });
 }, [interactiveState]);
```

Reply:
Nice -- thanks 👍 I've made that change. Look good?